common/ptr.h

```diff
@@ -366,6 +414,82 @@ class WeakPtr : public BasePtr<T> {
 	SharedPtr<T> lock() const {
 		return SharedPtr<T>(*this);
 	}
+
+	/**
+	 * Implicit conversion operator to bool for convenience, to make
+	 * checks like "if (sharedPtr) ..." possible.
+	 */
+	bool operator_bool() const {
+		return this->_pointer != nullptr;
+	}
+
+	/**
+	 * Returns whether the referenced object isn't valid
+	 */
+	bool expired() const {
+		return this->_tracker == nullptr || this->_tracker->getStrongCount() == 0;
+	}
+
+	WeakPtr<T> &operator=(const WeakPtr<T> &r) {
+		reset(r);
+		return *this;
+	}
+
+	template<class T2>
+	WeakPtr<T> &operator=(const WeakPtr<T2> &r) {
+		reset(r);
+		return *this;
+	}
+
+	WeakPtr<T> &operator=(const BasePtr<T> &r) {
+		reset(r);
+		return *this;
+	}
+
+	template<class T2>
+	WeakPtr<T> &operator=(const BasePtr<T2> &r) {
```

Comment:
Isn't this one enough for all cases?

Reply:
No, as with the copy constructor there needs to be an exact match for the assignment operator or the compiler will generate one that doesn't work correctly.